### PR TITLE
lib/*: Replace `posix-event` with independent libraries

### PR DESCRIPTION
**Pull request by andreittr**
### Description of changes

This changeset replaces the `posix-event` library with two independent libraries covering its functionality:
- `posix-poll`: epoll files, poll(), and select()
- `posix-eventfd`: eventfd files

`posix-poll` supports simultaneous polling of both newvfs files as well as legacy vfscore files.

A compatibility Kconfig entry now serves as a shortcut for consumers of `posix-event` to now select both replacement libraries instead.

This PR moves the following syscalls out of `posix-event`:
- `eventfd*`
- `(p)select`
- `(p)poll`
- `epoll_*`

Depends on the following PRs, as well as all of their dependencies:
- https://github.com/unikraft/unikraft/pull/1165
- https://github.com/unikraft/unikraft/pull/1168
- https://github.com/unikraft/unikraft/pull/1170
- https://github.com/unikraft/unikraft/pull/1157
- https://github.com/unikraft/unikraft/pull/1162
- https://github.com/unikraft/unikraft/pull/1156

Merged branch for testing @ https://github.com/andreittr/unikraft/tree/ttr/feat-vfs/event.

Part of the vfs rearch. Questions and comments regarding design, implementation, and documentation are more than welcome and will help make the new vfs layer more clear, intuitive, and performant.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A

### Additional configuration

All event polling should continue working as normal. In addition, simultaneous polling of both newvfs files (like eventfd) and legacy vfscore files should work as expected.

### CODE REVIEW

File: lib/posix-poll/epoll.c [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+{
+	struct uk_file *f;
+	unsigned int mode;
+	int ret;
+
+	if (unlikely(flags & ~EPOLL_CLOEXEC))
+		return -EINVAL;
+
+	f = uk_epollfile_create();
+	if (!f)
```
Comment (skuenzer) [inline lib/posix-poll/epoll.c]: We could put `unlikely`, here, too.

File: lib/posix-poll/select.c [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		   struct timespec *restrict timeout,
+		   const struct uk_ksigset *sigset)
+{
+	const struct uk_file *ef;
+	int ret;
+	int monitored;
+
+	if (unlikely(nfds < 0))
+		return -EINVAL;
+	if (timeout && uk_time_spec_to_nsec(timeout) < 0)
```
Comment (skuenzer) [inline lib/posix-poll/select.c]: `unlikely`?

File: lib/Config.uk [hunk 3]
```diff
@@ -32,3 +32,13 @@ config HAVE_SYSCALL
...
+
+##
+# Compatibility entries for legacy libraries
+##
+
+config LIBPOSIX_EVENT
+	bool
+	select LIBPOSIX_POLL
+	select LIBPOSIX_EVENTFD
+	default n
```
Comment (skuenzer) [inline lib/Config.uk]: In general `default n` we never need to type because that is the default.

Comment (andreittr) [reply lib/Config.uk]: Removed. A future PR should delete all those `default n` lines above (and in the rest of the Unikraft codebase, I don't think I've ever seen a config without a default n)

File: lib/posix-poll/epoll.c [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <errno.h>
+
+#include <uk/arch/atomic.h>
+#include <uk/assert.h>
+#include <uk/essentials.h>
+#include <uk/file-nops.h>
```
Comment (skuenzer) [inline lib/posix-poll/epoll.c:12]: `#include <uk/file/nops.h>`

File: lib/posix-eventfd/eventfd.c [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <sys/eventfd.h>
+
+#include <uk/alloc.h>
+#include <uk/errptr.h>
+#include <uk/file-nops.h>
```
Comment (skuenzer) [inline lib/posix-eventfd/eventfd.c:11]: `#include <uk/files/nops.h>`

**Comment by skuenzer** [comment]
Hi @andreittr, I rebased the PR and will push it to your branch.

**Review by skuenzer — APPROVED** [review]
Hi @andreittr, thanks a lot for your work. I will fix the two `#include` while upstreaming.

Reviewed-by: Simon Kuenzer <simon@unikraft.io>
Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request merged**